nemo/collections/asr/parts/mixins/mixins.py

```diff
@@ -45,65 +45,153 @@ def _setup_tokenizer(self, tokenizer_cfg: DictConfig):
 
             if 'hf_kwargs' in tokenizer_cfg:
                 with open_dict(self.cfg.tokenizer):
                     self.cfg.tokenizer.hf_kwargs = tokenizer_cfg.get('hf_kwargs')
 
-        if self.tokenizer_type not in ['bpe', 'wpe']:
+        if self.tokenizer_type not in ['bpe', 'wpe', 'agg']:
             raise ValueError(
-                "`tokenizer.type` must be either `bpe` for SentencePiece tokenizer or "
+                "`tokenizer.type` must be either `agg` for aggregate, `bpe` for SentencePiece tokenizer or "
                 "`wpe` for BERT based tokenizer"
             )
 
-        if self.tokenizer_type == 'bpe':
+        if self.tokenizer_type == 'agg':
+            logging.debug('_setup_tokenizer: detected an aggregate tokenizer')
+            # need to de-register any old artifacts
+            if hasattr(self, 'cfg'):
+                with open_dict(self.cfg):
+                    self.cfg.tokenizer.pop('model_path', None)
+                    self.cfg.tokenizer.pop('vocab_path', None)
+                    self.cfg.tokenizer.pop('spe_tokenizer_vocab', None)
+            if hasattr(self, 'artifacts'):
+                self.artifacts.pop('tokenizer.model_path', None)
+                self.artifacts.pop('tokenizer.vocab_path', None)
+                self.artifacts.pop('tokenizer.spe_tokenizer_vocab', None)
+                for akey in list(self.artifacts.keys()):
+                    if akey.startswith('tokenizer.tokenizers.'):
+                        self.artifacts.pop(akey)
+
+            tokenizers_dict = {}
+            for lang, tcfg in self.tokenizer_cfg['tokenizers'].items():
+                (tokenizer, model_path, vocab_path, spe_vocab_path,) = self._make_tokenizer(tcfg, lang)
+                tokenizers_dict[lang] = tokenizer
+                if hasattr(self, 'cfg'):
+                    with open_dict(self.cfg):
+                        self.cfg.tokenizer.tokenizers[lang]['dir'] = self.tokenizer_cfg['tokenizers'][lang]['dir']
+                        self.cfg.tokenizer.tokenizers[lang]['type'] = self.tokenizer_cfg['tokenizers'][lang]['type']
+
+                # pop these now that we don't need them
+                if "dir" in tcfg:
+                    self.tokenizer_cfg['tokenizers'][lang].pop('dir')
+                self.tokenizer_cfg['tokenizers'][lang].pop('type')
+                self.tokenizer_cfg['tokenizers'][lang].pop('hf_kwargs', {})
+
+            self.tokenizer = tokenizers.AggregateTokenizer(tokenizers_dict)
+        else:
+            logging.debug('_setup_tokenizer: detected a monolingual tokenizer')
+            # need to de-register any old artifacts
+            if hasattr(self, 'cfg'):
+                with open_dict(self.cfg):
+                    self.cfg.tokenizer.pop('tokenizers', None)
+            if hasattr(self, 'artifacts'):
+                for akey in list(self.artifacts.keys()):
+                    if akey.startswith('tokenizer.tokenizers.'):
```

Comment:
Use class variables for accessing such internal things. Also document why this is necessary.

Reply:
class var added.. comments added